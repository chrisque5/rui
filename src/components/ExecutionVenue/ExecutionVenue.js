import PropTypes from 'prop-types';
import { StarFilled } from '@ant-design/icons';
import {
  Form, Select, Tooltip,
} from 'antd';
import { ids } from '../../utils/constants';

const ExecutionVenue = ({
  isLoading, options, addPreferenceClick,
}) => (
  <Form.Item
    labelCol={{ span: 12 }}
    wrapperCol={{ span: 12 }}
    label={(
      <>
        <span className="execution-venue-label">Venue</span>
        <Tooltip
          title="Add Execution Venue to client manager"
          mouseEnterDelay={0.3}
          overlayStyle={
            { fontSize: '0.8em', padding: 0, margin: 0 }
          }
        >
          <StarFilled
            className="star-icon"
            data-testid="addPreference"
            onClick={addPreferenceClick}
          />
        </Tooltip>
      </>
    )}
    rules={[{ required: true, message: 'Execution venue required' }]}
    name={ids.EXECUTION_VENUE}
  >

    <Select loading={isLoading} data-testid="ddlExecutionVenue">
      {options && options.map(
        (executionVenue) => {
          const { venueId } = executionVenue;
          return (
            <Select.Option
              data-testid={`ddlExecutionVenue_${venueId}`}
              key={venueId}
              value={venueId}
            >
              {venueId}
            </Select.Option>
          );
        },
      )}
    </Select>
  </Form.Item>
);

ExecutionVenue.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  addPreferenceClick: PropTypes.func.isRequired,
};

export default ExecutionVenue;
