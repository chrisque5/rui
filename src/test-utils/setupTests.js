// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
