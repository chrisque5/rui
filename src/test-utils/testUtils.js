/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
