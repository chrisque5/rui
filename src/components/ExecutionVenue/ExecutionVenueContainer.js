import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal } from 'antd';
import ExecutionVenue from './ExecutionVenue';
import * as executionVenueActions from '../../actions/executionVenueActions';
import * as userActions from '../../actions/userActions';
import * as uiActions from '../../actions/uiActions';
import { ids, dealTypes } from '../../utils/constants';
import { filterSpotExecutionVenues } from '../../utils/helper';

class ExecutionVenueContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: true,
    };
  }

  componentDidMount = () => {
    this.props.executionVenueActions.loadExecutionVenues(this.props.dealType)
      .then(() => {
        this.setState((prevState) => ({
          ...prevState,
          isLoading: !prevState.isLoading,
        }));
      });
  }

  componentDidUpdate = () => {
    if (!this.state.isLoading) {
      this.setDefault();
    }
  }

  reduceExecutionVenues = () => {
    if (this.props.dealType === dealTypes.SPT) {
      return filterSpotExecutionVenues(this.props.executionVenues);
    }
    return this.props.executionVenues;
  }

  setDefault = () => {
    const { form } = this.props;
    const reducedExecVenues = this.reduceExecutionVenues();
    const storedVenue = form.getFieldValue(ids.EXECUTION_VENUE);

    if (!storedVenue && reducedExecVenues.length > 0) {
      const [defaultVenue = {}] = reducedExecVenues;
      const { venueId = undefined } = defaultVenue;
      form.setFieldsValue({ executionVenue: venueId });
    }
  }

  addToPreference = () => {
    const executionVenue = this.props.form.getFieldValue(`${ids.EXECUTION_VENUE}`);
    const selectedBrokerDeskId = this.props.userInterface.selectedPreferenceBroker;
    if (!selectedBrokerDeskId || selectedBrokerDeskId === '') {
      Modal.warning({
        title: 'Warning',
        content: 'No Broker tab open to add the favourite. Please open a Broker tab.',
      });
    } else if (executionVenue) {
      this.props.userActions.addExecutionVenuePreference(selectedBrokerDeskId, executionVenue);
    }
  };

  render() {
    return (
      <ExecutionVenue
        {...this.props}
        isLoading={this.state.isLoading}
        options={this.reduceExecutionVenues()}
        addPreferenceClick={this.addToPreference}
      />
    );
  }
}

ExecutionVenueContainer.propTypes = {
  executionVenueActions: PropTypes.shape().isRequired,
  executionVenues: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  userActions: PropTypes.shape().isRequired,
  uiActions: PropTypes.shape().isRequired,
  userInterface: PropTypes.shape().isRequired,
  form: PropTypes.shape().isRequired,
  dealType: PropTypes.string.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  executionVenueActions: bindActionCreators(executionVenueActions, dispatch),
  userActions: bindActionCreators(userActions, dispatch),
  uiActions: bindActionCreators(uiActions, dispatch),
});

const mapStateToProps = (state) => ({
  executionVenues: state.executionVenues,
  userInterface: state.ui,
});

export default connect(mapStateToProps, mapDispatchToProps)(ExecutionVenueContainer);
