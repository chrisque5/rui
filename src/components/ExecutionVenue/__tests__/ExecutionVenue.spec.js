import { renderWithForm, fireEvent, waitFor } from 'test-utils/testUtils';
import ExecutionVenue from '../ExecutionVenue';

const form = {};
const addPreferenceClick = jest.fn();
const options = [{
  id: null,
  lockSequence: null,
  venueId: 'TPSEF',
  productType: 'FX',
  venueType: 'SEF',
  reportingRegime: 'DoddFrankAct',
},
{
  id: null,
  lockSequence: null,
  venueId: 'XOFF',
  productType: 'FX',
  venueType: 'OffFacility',
  reportingRegime: 'DoddFrankAct',
}];

const defaultProps = {
  form,
  addPreferenceClick,
  options,
  isLoading: false,
};

describe('<ExecutionVenue/>', () => {
  test('It Renders ', () => {
    const { container } = renderWithForm(<ExecutionVenue {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('Add to favorites', async () => {
    const { getByTestId } = renderWithForm(<ExecutionVenue {...defaultProps} />);
    const star = await waitFor(() => getByTestId('addPreference'));
    fireEvent.click(star);
    expect(addPreferenceClick).toHaveBeenCalledTimes(1);
  });
});
