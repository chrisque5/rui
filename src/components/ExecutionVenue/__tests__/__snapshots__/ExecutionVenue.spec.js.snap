// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExecutionVenue/> It Renders  1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-col-12 ant-form-item-label"
      >
        <label
          class="ant-form-item-required"
          for="executionVenue"
          title=""
        >
          <span
            class="execution-venue-label"
          >
            Venue
          </span>
          <span
            aria-label="star"
            class="anticon anticon-star star-icon"
            data-testid="addPreference"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="star"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-col-12 ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-select ant-select-single ant-select-show-arrow"
              data-testid="ddlExecutionVenue"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="executionVenue_list_0"
                    aria-autocomplete="list"
                    aria-controls="executionVenue_list"
                    aria-haspopup="listbox"
                    aria-owns="executionVenue_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="executionVenue"
                    readonly=""
                    role="combobox"
                    style="opacity: 0;"
                    type="search"
                    unselectable="on"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-placeholder"
                />
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
